### PR TITLE
Add a Python unit test for combining DDP and Distributed Autograd/Optimizer

### LINKED ISSUE #33482: Distributed Data Parallel for models with remote parameter servers
## 🚀 Feature

The current [DDP] implementation (https://pytorch.org/tutorials/intermediate/ddp_tutorial.html) require a model is on a single node (maybe multiple devices). DDP won't work if a model has a remote parameter server and its `forward()` method makes RPCs to the remote parameter server. This feature is to make DDP work for models with parameter servers by leveraging the existing [Distributed Autograd/Optimizer](https://pytorch.org/docs/stable/notes/distributed_autograd.html).

Note that the parameter servers can't use DDP in this use case.

## Motivation

To support the following model

```
class SimpleNet(nn.Module):
    def __init__(self, d_in, d_out):
        super(SimpleNet, self).__init__()
        self.net = nn.Linear(d_in, d_out)
        self.relu = nn.ReLU()

    def forward(self, input):
        return self.relu(self.net(input))

class DdpModelWithRpc(nn.Module):
    def __init__(self, remote_parameter_server):
        super(DdpModelWithRpc, self).__init__()
        self.net1 = DDP(SimpleNet(5, 8))
        self.rps = remote_parameter_server
        self.rref = RPC call(remote_parameter_server, SimpleNet, 8, 5)
        self.net2 = DDP(SimpleNet(5, 3))

    def forward(self, x):
        x = self.net1(x)
        x = RPC(self.rps, SimpleNet.forward, x)
        return self.net2(x)

```

## Pitch

TDB

## Alternatives

TDB

## Additional context

CC: @pritamdamania87  @mrshenli @zhaojuanmao

### PR DESCRIPTION
**Pull request by hczhu**
This is a final version of #33580 
Summary:
This is to shape how a PyTorch user would use this new feature.
All test cases pass, though the backward() and gradient updates actually don't work correctly. The next step is to add correctness checks.

Test Plan 
```
$ python test/distributed/test_ddp_with_rpc.py

...

----------------------------------------------------------------------
Ran 6 tests in 6.762s

OK

```
#33482

### CODE REVIEW

File: test/distributed/test_ddp_with_rpc.py [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+        sparse_features=torch.zeros(n, dtype=torch.long),
+        labels=torch.zeros(n, dtype=torch.long),
+    )
+    idx = 0
+    for x in range(2):
+        for y in range(2):
+            for z in range(2):
+                training_examples.dense_features[idx, :] = torch.Tensor((x, y))
+                training_examples.sparse_features[idx] = z
+                training_examples.labels[idx] = x ^ y ^ z
```
Comment (mrshenli) [inline test/distributed/test_ddp_with_rpc.py]: is there any reason for these value assignments?

File: test/distributed/test_ddp_with_rpc.py [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+
+        gLogger.info(f"Loss: {loss}.")
+        return loss
+
+
+def get_training_examples():
+    n = 8
+    training_examples = FeatureSet(
+        dense_features=torch.zeros((n, D_DENSE)),
+        sparse_features=torch.zeros(n, dtype=torch.long),
```
Comment (mrshenli) [inline test/distributed/test_ddp_with_rpc.py]: are we going to use real sparse tensors for this?

File: test/distributed/test_ddp_with_rpc.py [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+        process_group_for_ddp = dist_c10d.new_group(ranks=TRAINER_RANKS)
+        gLogger.info(f"The remote worker is running.")
+        dist.destroy_process_group(process_group_for_ddp)
+        gLogger.info(f"Exiting remote worker.")
+
+    @dist_init
+    def _trainer_process(self, rank: int):
+        gLogger.info(f"Running the trainer #{rank}...")
+
+    def do_test_on_master(
```
Comment (mrshenli) [inline test/distributed/test_ddp_with_rpc.py]: since this is a test, can we assert on sth? e.g., output, grad, etc.

**Review by mrshenli — COMMENTED** [review]
1. IIUC, CI didn't launch this test, because we need to add this test file to `test/run_test.py`. You can search for `test_rpc_spawn`.
2. This is a super complex test, which would require rich documentations before and along with the implementation. 
3. What is your plan for landing this? Are we only going to land this after we finish this feature?

**Review by mrshenli — COMMENTED** [review]
Regarding the implementation plan, what @gqchen mentioned today makes sense to me too. We can

1. Enable post hooks in distributed autograd engine. This does not need to involve DDP and can be tested independently. Looks like we can use [`Tensor.register_hook`](https://pytorch.org/docs/stable/_modules/torch/tensor.html#Tensor.register_hook) install post hooks from Python (not 100% sure). So, the flow could be, create model, install hooks, run forward-backward, and then check if the hook is indeed launched. Cannot test with DDP in this stage yet, but DDP would look for grad in `Parameter.grad` field.
2. Add an API in `dist_autograd` to mark given parameters and then store their grad in both autograd context and in `Parameter.grad` using the same tensor storage. Add a smaller test to test this feature, e.g., the model can contain a local Linear layer and a remote Linear layer, where DDP wraps the local Linear layer which is connected to the remote Linear layer using RPC.
3. Land this more complex test.

**Comment by hczhu** [comment]
> 3. What is your plan for landing this? Are we only going to land this after we finish this feature?

My plan is to land this after finishing ddp+rpc feature.

File: test/distributed/test_ddp_with_rpc.py [hunk 4]
```diff
@@ -0,0 +1,443 @@
...
+
+class RemoteEM(nn.Module):
+    def __init__(self, num_embeddings: int, embedding_dim: int):
+        gLogger.info(f"Initing RemoteEM with {num_embeddings} {embedding_dim}")
+        super(RemoteEM, self).__init__()
+        init_em = [1] * embedding_dim
+        self.em = nn.EmbeddingBag(
+            num_embeddings,
+            embedding_dim,
+            _weight=torch.Tensor([init_em] * num_embeddings),
```
[the remark this answers is not in the document]
Comment (hczhu) [reply test/distributed/test_ddp_with_rpc.py]: Fixed

File: test/distributed/test_ddp_with_rpc.py [hunk 5]
```diff
@@ -0,0 +1,443 @@
...
+            self.fc1 = DistributedDataParallel(
+                self.fc1,
+                process_group=process_group_for_ddp,
+                check_reduction=True,
+            )
+            self.fc2 = DistributedDataParallel(
+                self.fc2,
+                process_group=process_group_for_ddp,
+                check_reduction=True,
+            )
```
[the remark this answers is not in the document]
Comment (hczhu) [reply test/distributed/test_ddp_with_rpc.py]: Yes, want to test the situation that there are a local net before RPC and one after RPC.

File: test/distributed/test_ddp_with_rpc.py [hunk 6]
```diff
@@ -0,0 +1,443 @@
...
+
+    def forward(self, input: FeatureSet):
+        gLogger.debug(f"Running HybridModel.forward on {input}")
+        sparse = _remote_method(
+            RemoteEM.forward, self.remote_em_rref, input.sparse_features
+        )
+        # The same size of mini batch.
+        assert sparse.shape[0] == input.dense_features.shape[0]
+        dense = self.fc1(input.dense_features)
+        x = torch.cat((dense, sparse), 1)
```
[the remark this answers is not in the document]
Comment (hczhu) [reply test/distributed/test_ddp_with_rpc.py]: I should rename it. It's not a sparse tensor.

File: test/distributed/test_ddp_with_rpc.py (hunk 2 above)
Comment (hczhu) [reply test/distributed/test_ddp_with_rpc.py]: No, it's sparse futures (e.g. user ids) instead of sparse tensors.

File: test/distributed/test_ddp_with_rpc.py (hunk 3 above)
Comment (hczhu) [reply test/distributed/test_ddp_with_rpc.py]: I'll add assertions later.

File: test/distributed/test_ddp_with_rpc.py (hunk 1 above)
Comment (hczhu) [reply test/distributed/test_ddp_with_rpc.py]: Not really. There are 8 examples as the 8 vertices of a cube in a 3-dimensional space. Each edge of the cube connects 2 vertices with different labels.

File: test/distributed/test_ddp_with_rpc.py [hunk 7]
```diff
@@ -0,0 +1,443 @@
...
+                check_reduction=True,
+            )
+
+        gLogger.info(
+            f"HybridModel has {len(list(self.parameters()))} groups of parameters."
+        )
+
+    def forward(self, input: FeatureSet):
+        gLogger.debug(f"Running HybridModel.forward on {input}")
+        sparse = _remote_method(
```
[the remark this answers is not in the document]
Comment (hczhu) [reply test/distributed/test_ddp_with_rpc.py]: No, I'll rename the varialbe.

File: test/distributed/test_ddp_with_rpc.py [hunk 8]
```diff
@@ -0,0 +1,443 @@
...
+class Trainer:
+    def __init__(
+        self,
+        remote_em_rref: rpc.RRef,
+        remote_net_rref: rpc.RRef,
+        ddp_mode: DdpMode,
+        rank: int,
+    ):
+        gLogger.info(
+            f"Initing trainer process group by traner #{rank} with ranks {TRAINER_RANKS}"
```
[the remark this answers is not in the document]
Comment (hczhu) [reply test/distributed/test_ddp_with_rpc.py]: done.

**Comment by hczhu** [comment]
> 1. Enable post hooks in distributed autograd engine. This does not need to involve DDP and can be tested independently. Looks like we can use [`Tensor.register_hook`](https://pytorch.org/docs/stable/_modules/torch/tensor.html#Tensor.register_hook) install post hooks from Python (not 100% sure). So, the flow could be, create model, install hooks, run forward-backward, and then check if the hook is indeed launched. Cannot test with DDP in this stage yet, but DDP would look for grad in `Parameter.grad` field.

@mrshenli 
Just FYI, `Tensor.register_hook` registers a pre-hook, while DDP installs post-hooks.

File: test/distributed/test_ddp_with_rpc.py [hunk 9]
```diff
@@ -0,0 +1,444 @@
...
+                    )
+                )
+            log_loss = 0
+            for future in futures:
+                log_loss += future.wait()
+            gLogger.info(
+                f"Log loss at epoch #{epoch}: {log_loss / NUM_TRAINERS}"
+            )
+
+    @dist_init
```
Comment (pritamdamania87) [inline test/distributed/test_ddp_with_rpc.py]: Shouldn't dist_init wrap the individual test functions instead of these methods?

File: test/distributed/test_ddp_with_rpc.py [hunk 10]
```diff
@@ -0,0 +1,444 @@
...
+
+    def tearDown(self):
+        super(TestDdpWithRpc, self).tearDown()
+
+    @dist_init
+    def _remote_worker_process(self):
+        process_group_for_ddp = dist_c10d.new_group(ranks=TRAINER_RANKS)
+        gLogger.info(f"The remote worker is running.")
+        dist.destroy_process_group(process_group_for_ddp)
+        gLogger.info(f"Exiting remote worker.")
```
Comment (pritamdamania87) [inline test/distributed/test_ddp_with_rpc.py]: Wouldn't the remote worker process exit before the master/trainer has a chance to run some computations on it? We should add a `dist.barrier()` at the end of `_do_test` to ensure all processes complete execution before returning.

I'm guessing right now rpc.shutdown() in `dist_init` is cleanly doing this, but its probably better to make this explicit in the unit test.

File: test/distributed/test_ddp_with_rpc.py [hunk 11]
```diff
@@ -0,0 +1,444 @@
...
+
+    @dist_init
+    def _trainer_process(self, rank: int):
+        gLogger.info(f"Running the trainer #{rank}...")
```
Comment (pritamdamania87) [inline test/distributed/test_ddp_with_rpc.py]: Same as above, wouldn't trainer process exit before we have a chance to run some computations?

File: test/distributed/test_ddp_with_rpc.py [hunk 12]
```diff
@@ -0,0 +1,444 @@
...
+class Trainer:
+    def __init__(
+        self,
+        remote_em_rref: rpc.RRef,
+        remote_net_rref: rpc.RRef,
+        ddp_mode: DdpMode,
+        rank: int,
+    ):
+        gLogger.info(
+            f"Initializing trainer process group by traner #{rank} with ranks {TRAINER_RANKS}"
```
Comment (mmphego) [inline test/distributed/test_ddp_with_rpc.py]: Typo fix.
```suggestion             
            f"Initializing trainer process group by trainer #{rank} with ranks {TRAINER_RANKS}"
```

File: test/distributed/test_ddp_with_rpc.py [hunk 13]
```diff
@@ -0,0 +1,444 @@
...
+        self.remote_net_rref = remote_net_rref
+        self.hybrid_module = HybridModel(
+            self.remote_em_rref,
+            self.remote_net_rref,
+            self.process_group_for_ddp
+            if ddp_mode in (DdpMode.INSIDE,)
+            else None,
+        )
+        if ddp_mode == DdpMode.OUTSIDE:
+            gLogger.info(f"Wrapping the whole hybride module into DDP.")
```
Comment (mmphego) [inline test/distributed/test_ddp_with_rpc.py]: I think you meant to write *hybrid* instead of *hybride*.

```suggestion
            gLogger.info(f"Wrapping the whole hybrid module into DDP.")
```

File: test/distributed/test_ddp_with_rpc.py (hunk 10 above)
Comment (hczhu) [reply test/distributed/test_ddp_with_rpc.py]: > I'm guessing right now rpc.shutdown() in dist_init is cleanly doing this, but its probably better to make this explicit in the unit test.

That's correct.

**Review by pritamdamania87 — CHANGES_REQUESTED** [review]
Can we update this PR with appropriate assertions and verify all unit tests pass with https://github.com/pytorch/pytorch/pull/34501?

**Comment by hczhu** [comment]
Merged this PR to #37998
